### PR TITLE
Modify Study Builder to recreate workflow from 

### PR DESCRIPTION
**Pull request by MocanaAtBroad**
## Context
Tweaked study builder to add --only-update-workflow. Will delete existing workflow configuration in database and reload it from study.conf file

## FUD Score

_Overall, how are you feeling about these changes?_

- [ X] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [X ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ X] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/WorkflowDao.java [hunk 1]
```diff
@@ -64,6 +64,13 @@ default void insertTransitions(List<WorkflowTransition> transitions) {
         }
     }
 
+    default int deleteStudyWorkflow(long studyId) {
```
Comment (evgeniipr) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/WorkflowDao.java:67]: Is this method called from `TransactionWrapper.useTxn`?  As it calls several update queries.

Comment (evgeniipr) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/WorkflowDao.java:67]: Ah, yes it is. Just checked from code.

File: study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/StudyBuilderCli.java [hunk 2]
```diff
@@ -146,6 +147,11 @@ private void run(String[] args) throws Exception {
             execute(builder::runWorkflow, isDryRun);
             log("done");
             return;
+        } else if (cmd.hasOption("only-update-workflow")) {
+            log("executing workflow setup...");
```
Comment (evgeniipr) [inline study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/StudyBuilderCli.java]: Should it be a bit different log message comparing to 'only-workflow' option?

**Review by evgeniipr — DISMISSED** [review]

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiWorkflowTransition.java [hunk 3]
```diff
@@ -22,4 +25,10 @@ long insertActive(@Bind("studyId") long studyId,
 
     @SqlUpdate("delete from workflow_transition where workflow_transition_id = :transitionId")
     int deleteById(@Bind("transitionId") long transitionId);
+
+    @SqlUpdate("delete from workflow_transition where umbrella_study_id=:studyId")
+    int deleteAllForStudy(@Bind("studyId") long studyId);
+
+    @SqlQuery("select precondition_expression_id from workflow_transition where umbrella_study_id=:studyId")
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiWorkflowTransition.java]: Precondition expression is optional, so maybe add in a where-clause?
```
and precondition_expression_id is not null
```

File: study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/StudyBuilderCli.java [hunk 4]
```diff
@@ -58,6 +58,7 @@ private void run(String[] args) throws Exception {
         options.addOption(null, "dry-run", false, "run study setup or custom task without saving");
         options.addOption(null, "only-activity", true, "only run activity setup for given activity code");
         options.addOption(null, "only-workflow", false, "only run workflow setup");
+        options.addOption(null, "only-update-workflow", false, "recreate workflow from configuration file");
```
Comment (yufengwng) [inline study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/StudyBuilderCli.java]: Maybe call the new flag `--recreate-workflow`? "update" makes me think that it'll selectively add new workflows or update current ones.

File: study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/WorkflowBuilder.java [hunk 5]
```diff
@@ -29,13 +29,23 @@ public WorkflowBuilder(Config cfg, StudyDto studyDto) {
...
+        insertWorkflow(handle);
     }
 
-    private void insertTransitions(Handle handle) {
+    void runUpdate(Handle handle) {
         if (!cfg.hasPath("workflowTransitions")) {
             return;
         }
+        int deletionCount = handle.attach(WorkflowDao.class).deleteStudyWorkflow(studyDto.getId());
+        LOG.info("Deleted {} workflow transitions PEX expressions", deletionCount);
```
Comment (yufengwng) [inline study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/WorkflowBuilder.java]: The count we get back is actually for the number of transitions, so the log message here is a bit misleading.

**Review by yufengwng — COMMENTED** [review]
A few small comments, but overall looks good!

File: study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/StudyBuilderCli.java (hunk 4 above)
Comment (MocanaAtBroad) [reply study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/StudyBuilderCli.java]: Sure. 0a62f2fa

File: study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/WorkflowBuilder.java (hunk 5 above)
Comment (MocanaAtBroad) [reply study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/WorkflowBuilder.java]: RIght. 6398589b

File: study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/StudyBuilderCli.java [hunk 6]
```diff
@@ -146,6 +149,15 @@ private void run(String[] args) throws Exception {
             execute(builder::runWorkflow, isDryRun);
             log("done");
             return;
+        } else if (cmd.hasOption("only-recreate-workflow")) {
+            if (studyHasPatch(cfgPath) && !promptForConfirmation("There are patches for this study. Have you checked that workflow DOES "
+                    + "NOT refer to any activities added in patches for this study?")) {
+                return;
+            }
```
Comment (MocanaAtBroad) [inline study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/StudyBuilderCli.java:156]: Based on Slack exchange with @yufengwng , just adding a confirmation step. Guess we won't be able to use this option every time if there are activities created in patches that are referenced by workflow, but can be used sometimes, like for Osteo.

**Review by yufengwng — APPROVED** [review]

**Pull request merged**